File: projects/ancestor/graph.py
"""
Simple graph implementation
"""
from util import Stack, Queue  # These may come in handy

class Graph:

    """Represent a graph as a dictionary of vertices mapping labels to edges."""
    def __init__(self):
        self.vertices = {}

    def add_vertex(self, vertex_id):
        """
        Add a vertex to the graph.
        """
        self.vertices[vertex_id] = set()

    def add_edge(self, v1, v2):
        """
        Add a directed edge to the graph.
        """
        if v1 in self.vertices and v2 in self.vertices:
            self.vertices[v1].add(v2)
        else:
            raise IndexError("Vertex does not exist")

    def get_neighbors(self, vertex_id):
        """
        Get all neighbors (edges) of a vertex.
        """
        return self.vertices[vertex_id]

    def bft(self, starting_vertex):
        """
        Print each vertex in breadth-first order
        beginning from starting_vertex.
        """
        queue = Queue()
        visited = set()

        queue.enqueue(starting_vertex)

        while queue.size() > 0:
            vertex = queue.dequeue()

            if vertex in visited:
                continue
            else:
                visited.add(vertex)

            print(vertex)

            for neighbor in self.get_neighbors(vertex):
                queue.enqueue(neighbor)
            

    def dft(self, starting_vertex):
        """
        Print each vertex in depth-first order
        beginning from starting_vertex.
        """
        stack = Stack()
        visited = set()

        stack.push(starting_vertex)

        while stack.size() > 0:
            vertex = stack.pop()

            if vertex in visited:
                continue
            else:
                visited.add(vertex)

            print(vertex)

            for neighbor in self.get_neighbors(vertex):
                stack.push(neighbor) 


    def dft_recursive(self, starting_vertex):
        """
        Print each vertex in depth-first order
        beginning from starting_vertex.

        This should be done using recursion.
        """

        visited = set()

        def dft_inner(vertex):
            if vertex in visited:
                return
            else:
                visited.add(vertex)
            
            print(vertex)

            for neighbor in self.get_neighbors(vertex):
                dft_inner(neighbor)

        dft_inner(starting_vertex)


    def bfs(self, starting_vertex, destination_vertex):
        """
        Return a list containing the shortest path from
        starting_vertex to destination_vertex in
        breath-first order.
        """
        visited = set()
        path_queue = Queue() 

        path_queue.enqueue([starting_vertex])

        while path_queue.size() > 0:
            path = path_queue.dequeue()
            last_vertex = path[-1]

            if last_vertex in visited:
                continue
            else:
                visited.add(last_vertex)

            for neighbor in self.get_neighbors(last_vertex):
                next_path = path.copy()
                next_path.append(neighbor)

                if neighbor == destination_vertex:
                    return next_path
                path_queue.enqueue(next_path)

    def dfs(self, starting_vertex, destination_vertex):
        """
        Return a list containing a path from
        starting_vertex to destination_vertex in
        depth-first order.
        """
        visited = {}
        to_search = Stack()
        to_search.push((starting_vertex, None))
        while to_search.size() > 0:
            (vertex, prev) = to_search.pop()
            if vertex not in visited:
                visited[vertex] = prev
                if vertex == destination_vertex:
                    step = vertex
                    path = []
                    while step is not None:
                        path.append(step)
                        step = visited[step]
                    return path[::-1]
            for edge in self.vertices[vertex]:
                if edge not in visited:
                    to_search.push((edge, vertex))

    def dfs_recursive(self, starting_vertex, destination_vertex):
        """
        Return a list containing a path from
        starting_vertex to destination_vertex in
        depth-first order.

        This should be done using recursion.
        """
        visited = set()

        def dft_inner(path):
            last_vertex = path[-1]
            
            if last_vertex in visited:
                return None
            else:
                visited.add(last_vertex)

            if last_vertex == destination_vertex:
                return path
            
            for neighbor in self.get_neighbors(last_vertex):
                next_path = path.copy()
                next_path.append(neighbor)
            
                found = dft_inner(next_path)
                if found:
                    return found
            
            return None

        return dft_inner([starting_vertex])

if __name__ == '__main__':
    graph = Graph()  # Instantiate your graph
    # https://github.com/LambdaSchool/Graphs/blob/master/objectives/breadth-first-search/img/bfs-visit-order.png
    graph.add_vertex(1)
    graph.add_vertex(2)
    graph.add_vertex(3)
    graph.add_vertex(4)
    graph.add_vertex(5)
    graph.add_vertex(6)
    graph.add_vertex(7)
    graph.add_edge(5, 3)
    graph.add_edge(6, 3)
    graph.add_edge(7, 1)
    graph.add_edge(4, 7)
    graph.add_edge(1, 2)
    graph.add_edge(7, 6)
    graph.add_edge(2, 4)
    graph.add_edge(3, 5)
    graph.add_edge(2, 3)
    graph.add_edge(4, 6)

    '''
    Should print:
        {1: {2}, 2: {3, 4}, 3: {5}, 4: {6, 7}, 5: {3}, 6: {3}, 7: {1, 6}}
    '''
    print(graph.vertices)

    '''
    Valid BFT paths:
        1, 2, 3, 4, 5, 6, 7
        1, 2, 3, 4, 5, 7, 6
        1, 2, 3, 4, 6, 7, 5
        1, 2, 3, 4, 6, 5, 7
        1, 2, 3, 4, 7, 6, 5
        1, 2, 3, 4, 7, 5, 6
        1, 2, 4, 3, 5, 6, 7
        1, 2, 4, 3, 5, 7, 6
        1, 2, 4, 3, 6, 7, 5
        1, 2, 4, 3, 6, 5, 7
        1, 2, 4, 3, 7, 6, 5
        1, 2, 4, 3, 7, 5, 6
    '''
    graph.bft(1)

    '''
    Valid DFT paths:
        1, 2, 3, 5, 4, 6, 7
        1, 2, 3, 5, 4, 7, 6
        1, 2, 4, 7, 6, 3, 5
        1, 2, 4, 6, 3, 5, 7
    '''
    graph.dft(1)
    graph.dft_recursive(1)

    '''
    Valid BFS path:
        [1, 2, 4, 6]
    '''
    print(graph.bfs(1, 6))

    '''
    Valid DFS paths:
        [1, 2, 4, 6]
        [1, 2, 4, 7, 6]
    '''
    print(graph.dfs(1, 6))
    print(graph.dfs_recursive(1, 6))

File: projects/ancestor/graph.js
"""
Simple graph implementation
"""
from util import Stack, Queue  # These may come in handy

class Graph:

    """Represent a graph as a dictionary of vertices mapping labels to edges."""
    def __init__(self):
        self.vertices = {}

    def add_vertex(self, vertex_id):
        """
        Add a vertex to the graph.
        """
        self.vertices[vertex_id] = set()

    def add_edge(self, v1, v2):
        """
        Add a directed edge to the graph.
        """
        if v1 in self.vertices and v2 in self.vertices:
            self.vertices[v1].add(v2)
        else:
            raise IndexError('Vertex does not exist')

    def get_neighbors(self, vertex_id):
        """
        Get all neighbors (edges) of a vertex.
        """
        return self.vertices[vertex_id]

    def bft(self, starting_vertex):
        """
        Print each vertex in breadth-first order
        beginning from starting_vertex.
        """
        # create empty queue
        queue = Queue()
        queue.enqueue(starting_vertex)
        # create set to store vertices
        visited = set()
        # while que isn't empty,
        while queue.size() > 0:
            dequeued = queue.dequeue()
            # if vertex not visitied, mark it as visitied then add neighbors
            if dequeued not in visited:
                visited.add(dequeued)
                for i in self.vertices[dequeued]:
                    queue.enqueue(i)

    def dft(self, starting_vertex):
        """
        Print each vertex in depth-first order
        beginning from starting_vertex.
        """
        # create empty stack
        s = Stack()
        s.push(starting_vertex)
        # create set to store vertices
        visited = set()
        # while stack isn't empty,
        while s.size() > 0:
            popped = s.pop()
            #if vertex not visited, mark it as visited then add neighbors
            if popped not in visited:
                visited.add(popped)
                for i in self.vertices[popped]:
                    s.push(i)

    def dft_recursive(self, starting_vertex, visited=None):
        """
        Print each vertex in depth-first order
        beginning from starting_vertex.
        This should be done using recursion.
        """
        if visited is None:
            visited = set() 
        visited.add(starting_vertex)
        for neighbor_vert in self.get_neighbors(starting_vertex):
            if neighbor_vert not in visited:
                self.dft_recursive(neighbor_vert, visited)

    def bfs(self, starting_vertex, destination_vertex):
        """
        Return a list containing the shortest path from
        starting_vertex to destination_vertex in
        breath-first order.
        """
        # create empty queue
        queue = Queue()
        starting = [starting_vertex]
        queue.enqueue(starting)
        # create set to store vertices
        visited = set()
        # while que isn't empty,
        while queue.size() > 0:
            path = queue.dequeue()
            vert = path[-1]
            # if vertex not visitied, mark it as visitied then add neighbors
            if vert not in visited:
                # check for destination target
                if vert == destination_vertex:
                  return path
                visited.add(vert)
                for neighbor_vert in self.vertices[vert]:
                  copy = path[:]
                  copy.append(neighbor_vert)
                  queue.enqueue(copy)

    def dfs(self, starting_vertex, destination_vertex):
        """
        Return a list containing a path from
        starting_vertex to destination_vertex in
        depth-first order.
        """
        # create empty stack
        s = Stack()
        starting = [starting_vertex]
        s.push(starting)
        # create set to store vertices
        visited = set()
        while s.size() > 0:
            path = s.pop()
            vert = path[-1]
            #if vertex not visited, mark it as visited then add neighbors
            if vert not in visited:
                # check for destination target
                if vert == destination_vertex:
                  return path
                visited.add(vert)
                for neighbor_vert in self.vertices[vert]:
                  copy = path[:]
                  copy.append(neighbor_vert)
                  s.push(copy)

    def dfs_recursive(self, starting_vertex, destination_vertex, visited=None, path=None):
        """
        Return a list containing a path from
        starting_vertex to destination_vertex in
        depth-first order.
        This should be done using recursion.
        """
        if visited is None:
            visited = set()
        if path is None:
            path = []
        visited.add(starting_vertex)
        path = path + [starting_vertex]
        if starting_vertex == destination_vertex:
            return path
        for neighbor_vert in self.get_neighbors(starting_vertex):
            if neighbor_vert not in visited:
                new = self.dfs_recursive(neighbor_vert, destination_vertex, visited, path)
                if new is not None:
                    return new
        return None